language/move-prover/bytecode/src/access_path.rs

```diff
@@ -0,0 +1,681 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    dataflow_analysis::{AbstractDomain, SetDomain},
+    stackless_bytecode::TempIndex,
+};
+use move_core_types::language_storage::StructTag;
+use move_model::{
+    model::{FunctionEnv, GlobalEnv, ModuleId, QualifiedId, StructId},
+    ty::{PrimitiveType, Type, TypeDisplayContext},
+};
+use num::BigUint;
+use std::{fmt, fmt::Formatter};
+
+type Address = BigUint;
+
+/// This file contains an abstraction of concrete *access paths*, which are canonical names for a particular cell in memory. Some examples of concrete paths are:
+/// - `0x7/M::T/f` (i.e., the field `f` of the `M::T` resource stored at address `0x7`
+/// - `Formal(0)/[2]` (i.e., the value stored at index 2 of the array bound the 0th formal of the current procedure)
+/// An abstract path is similar; it consists of the following components:
+/// - A *root*, which is either an abstract address or a local
+/// - Zero or more *offsets*, where an offset is a field, an unknown vector index, or an abstract struct type
+///
+/// Abstract addresses are a set containing constants and abstract access paths read from the environment. For example, in the following code:
+///
+/// struct S { f: u64 }
+///
+/// fun foo(x: S) {
+///   let a = if (*) { 0x1 } else { *&x.f }
+///    ... // program point 1
+/// }
+///
+///, the value of `a` will be `{ 0x1, Footprint(x/f) }` at program point 1.
+
+/// Fully qualified type identifier `base` bound to type actuals `types`
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AbsStructType {
+    /// Module ID and struct ID
+    base: QualifiedId<StructId>,
+    /// Instantiation of generic type parameters
+    types: Vec<Type>,
+}
+
+impl AbsStructType {
+    pub fn new(mid: &ModuleId, sid: StructId, types: Vec<Type>) -> Self {
+        AbsStructType {
+            base: mid.qualified(sid),
+            types,
+        }
+    }
+
+    pub fn get_type(&self) -> Type {
+        Type::Struct(self.base.module_id, self.base.id, self.types.clone())
+    }
+
+    /// Substitue the open types in self.types with caller `type_actuals`
+    pub fn substitute_footprint(&mut self, type_actuals: &[Type]) {
+        for t in self.types.iter_mut() {
+            *t = t.instantiate(type_actuals)
+        }
+    }
+
+    /// Returns a normalized representation of this type if it closed,
+    /// None if it is open
+    pub fn normalize(&self, env: &GlobalEnv) -> Option<StructTag> {
+        self.get_type().into_struct_tag(env)
+    }
+
+    pub fn is_closed(&self) -> bool {
+        for t in &self.types {
+            if t.is_open() {
+                return false;
+            }
+        }
+        true
+    }
+
+    pub fn display<'a>(&'a self, env: &'a GlobalEnv) -> AbsStructTypeDisplay<'a> {
+        AbsStructTypeDisplay { s: self, env }
+    }
+}
+
+pub struct AbsStructTypeDisplay<'a> {
+    s: &'a AbsStructType,
+    env: &'a GlobalEnv,
+}
+
+impl<'a> fmt::Display for AbsStructTypeDisplay<'a> {
```

Comment:
Nit: I would move the fmt::Display implementation to the end of the file. They are boilerplate distracting from understanding the data structures. Perhaps also all impl S after the data model has been introduced.